### PR TITLE
CLI-less webpack build

### LINKED ISSUE #1098: [Proposal] CLI-less webpack build
**I'm submitting a feature request (proposal)**

* **Library Version:**
aurelia-cli v1.0.0-beta.15

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Node Version:**
11.14.0

* **NPM Version:**
6.9.0

* **Browser:**
NA

* **Language:**
NA

* **Loader/bundler:**
Webpack

**Current behavior:**

The current build and run pipeline for webpack projects looks a bit overkill.

* **What is the expected behavior?**

It can be made a lot easier if the layer offered by aurelia-cli tasks are removed, and the default webpack and webpack-dev-server CLI do their job.

* **What is the motivation / use case for changing the behavior?**

The idea to have a aurelia-cli-less webpack build for a cli-genrated project was introduced to me [here](https://github.com/aurelia/cli/pull/1088#issuecomment-491507523). It makes less clutter, gives more power to the developer, and lesser things to maintain for Aurelia team.

I have prepared the following "thing" to propose the solution (and mostly as a note to me to understand `au build`, and `au run`, as I am not a heavy user of aurelia-cli).

---------------------------

#### TL;DR
No surprise that it can be done easily using npm scripts and webpack CLI parameters.

In order to have a cli-less webpack build, and run for the webpack users, it is first needed to evaluate what au-cli currently offers. At present au-cli offers 2 commands `au build` (utilizes `webpack`) and `au run` (utilizes `webpack-dev-server`). Following sections explore these 2.

#### `au build`

The `webpack.config.js` exports a function that creates the config based on the supplied arguments.

```javascript
module.exports = ({ production, server, extractCss, coverage, analyze, karma } = {}) =>({
    // webpack config for magic
})
```

`au build` via `buid.ts` script collects the values for these arguments from CLI arguments, and `aurelia.json` and uses Webpack API to build the app. Hypothetically considering, if that is the only function of `build.ts`, it can be simply replaced with the following command line (refer https://github.com/webpack/webpack/issues/2254#issuecomment-203528744).

```bash
webpack --env.production --env.server --env.extractCss --env.coverage --env.analyze --env.karma
```

##### Reducing the number of arguments for `webpack.config`

From a very naive perspective, there are 3 different setups: 

- development, 
- production, and 
- tests. 

From this perspective, the `karma`, and `coverage` parameters can be replaced with a single parameter `tests`.

I am not sure about the utility of the `server` parameter. To me it looks like it is used only during development as a metadata for `HtmlWebpackPlugin`. However, it is not used in `index.ejs`. Another usage of this is the conditional application of `CopyWebpackPlugin` during production or development build. This can also be replaced with the new `!tests` parameter.

The other two parameters `extractCss`, and `analyze` can be kept untouched.

With this, the resultant npm scripts might look something like this.
```json
scripts:{
    ...
    "build": "webpack --env.production --env.extractCss",
    "build:dev": "webpack --env.extractCss",
    "analyze": "webpack --env.analyze",
    "test": "au test"
    ...
}
```

##### Removing other sources of configuration

The 3 CLI arguments utilized by the script are `analyze`, `env`, and `watch`, all of which are replaceable using simply the webpack CLI arguments (as shown above).

Another source of build parameters is from `aurelia.json#build#options`

```json
"build": {
    "options": {
        "server": "dev",
        "extractCss": "prod",
        "coverage": false
    }
},
```

Based on this configuration, the build script decides whether to apply this configuration for current environment, that is `production`, or `development`. Again all these options can be de/activated using npm scripts. CLI can generate a recommended script for each of this environment. If the need be the app developer can always trivially change the parameters as per need.

The other tasks, such as deleting the previous distributions and configuring environment, can be left as it is as a gulp task, which is invoked as pre-task (either by `pre` hook or using [npm-run-all](https://www.npmjs.com/package/npm-run-all)).

#### `au run`

This is also very similar with `au build`, in terms of that it too collects the configurations from CLI, `aurelia.json`, and `webpack.config.js` and runs the app using `webpack-dev-server` with accumulated configuration options.

All the CLI options for the `au run` can be replaced with npm scripts as shown above. And the recommended defaults can directly be baked into the generated `webpack.config.js`. 

Following code fragment is not really needed for HMR or reloading the app. I tried without both the options (for HMR I referred the [webpack guide](https://webpack.js.org/guides/hot-module-replacement)), and it worked as expected, both with or w/o HMR. I think these are some old artifacts.

```typescript
  if (project.platform.hmr || CLIOptions.hasFlag('hmr')) {
    config.plugins.push(new webpack.HotModuleReplacementPlugin());
    config.entry.app.unshift('webpack-dev-server/client', 'webpack/hot/dev-server');
  } else {
    // removed "<script src="/webpack-dev-server.js"></script>" from index.ejs in favour of this method
    config.entry.app.unshift('webpack-dev-server/client');
  }
```

#### POC

For POC, please refer default-ts-webpack-app in [this repo](https://github.com/Sayan751/aurelia-cli-projects) which uses the webpack and webpack-dev-server cli solely to build and run the cli-generated default TS-Webpack app. The README shows the modified scripts. For specific details refer https://github.com/Sayan751/aurelia-cli-projects/commit/af008efc6330a03bf73aab3ffcd8174247d57f1b#diff-782ee31feceb525ffde151015edd994a.

**Note1:** My analysis here is solely based on the cli-generated default TS-Webpack app.

**Note2:** The user acceptance needs to be considered. If there is a majority of users accustomed with `au ` flavored tasks, it might not be a welcoming change.

Lastly, sorry for the wall of text :)

### PR DESCRIPTION
**Pull request by Sayan751**
This PR resolves #1098.

Summary of changes:

- Parameters consolidations and other minor changes in webpack.config.js. Removed the limiting options from ts-loader as it prevents compiling `*.spec.ts` files during test with karma(+webpack).
- Replaced the au build and run scripts with npm scripts. Affected areas: `package.json`, `aurelia_project/tasks`, and release check tests. For pre-build cleanup activity, there is a small gulpfile introduced.
- Changed `tsconfig.json` to have `typeroots` applied for all cases, without that the release checks were failing for couple of suites. Other changes here enable loading a json file with ES6 import syntax.
- Lastly converted the `environment.ts`s to json files, which are loaded using the [app-settings-loader](https://github.com/Sayan751/app-settings-loader).

### CODE REVIEW

File: skeleton/common/tsconfig.json__if_typescript [hunk 1]
```diff
@@ -44,6 +42,8 @@
     "lib": ["es2015", "dom"],
     "moduleResolution": "node",
     "baseUrl": "src",
+    "resolveJsonModule": true,
+    "esModuleInterop": true,
```
Comment (3cp) [inline skeleton/common/tsconfig.json__if_typescript]: I like `"esModuleInterop": true,` personally, but majority of Aurelia core team do not like this switch. I don't think you can pass review :-)

File: build/tasks/release-checks/tests/generic/au-cypress.js [hunk 2]
```diff
@@ -30,7 +30,7 @@ class AuCypressRunsTests extends Test {
   }
 
   execute() {
-    this.executeCommand = new ExecuteCommand('au', ['run'], (msg) => this.onOutput(msg));
+    this.executeCommand = new ExecuteCommand('npm', ['start'], (msg) => this.onOutput(msg));
```
Comment (3cp) [inline build/tasks/release-checks/tests/generic/au-cypress.js]: Since you normalised `au` commands to behind same npm scripts, can you update all the file names from `au-xxx.js` to maybe `xxx.js` because `au-` is misleading for the new webpack setup.

File: skeleton/webpack/webpack.config.js [hunk 3]
```diff
@@ -278,20 +279,23 @@ module.exports = ({ production, server, extractCss, coverage, analyze, karma } =
       // @if feat.babel
       {
         test: /\.js$/i, loader: 'babel-loader', exclude: nodeModulesDir,
-        options: coverage ? { sourceMap: 'inline', plugins: ['istanbul'] } : {}
+        options: tests ? { sourceMap: 'inline', plugins: ['istanbul'] } : {}
       },
       // @endif
       // @if feat.typescript
-      { test: /\.ts$/, loader: "ts-loader", options: { reportFiles: [ srcDir+'/**/*.ts'] }, include: srcDir },
```
Comment (3cp) [inline skeleton/webpack/webpack.config.js]: I think you accidentally removed this line from master.

File: skeleton/common/tsconfig.json__if_typescript [hunk 4]
```diff
@@ -32,10 +32,8 @@
     "types": ["node", "jest"],
     // @endif
 
-    // @if feat.webpack && feat['postcss-typical'] && feat.protractor
```
Comment (3cp) [inline skeleton/common/tsconfig.json__if_typescript]: I think you accidentally removed some lines?

Comment (3cp) [reply skeleton/common/tsconfig.json__if_typescript]: Ok, I saw your commit message. But we did this conditional typeRoots to do minimum impact for users of lerna (where `"./node_modules/@types"` does not work).

File: skeleton/common/tsconfig.json__if_typescript (hunk 1 above)
Comment (fkleuver) [reply skeleton/common/tsconfig.json__if_typescript]: Bad switch. :)

File: skeleton/webpack/webpack.config.js (hunk 3 above)
Comment (3cp) [reply skeleton/webpack/webpack.config.js]: Ok, saw your commit message for this one too. I guess the side effect is that you have to turn on typeRoots unconditionally, which is not idea.

**Comment by 3cp** [comment]
Sidenote: this is technically a breaking change for webpack app. But it's practically not a breaking change, because it doesn't affect/break existing webpack apps. This only affects newly created webpack app.

File: skeleton/common/tsconfig.json__if_typescript (hunk 1 above)
Comment (3cp) [reply skeleton/common/tsconfig.json__if_typescript]: @Sayan751 fight or run? :-)

Comment (Sayan751) [reply skeleton/common/tsconfig.json__if_typescript]: I prefer this as it let me import json as a module, and I can work with this object as a normal JS object. For readonly config, or translation files (i18n) it is quite handy. I am using it for quite a some time now. Till now, have not faced any issue. Would be great if you have some use cases where this can be problematic. This info will be very important to me personally to evaluate the usages in my project 😄

File: build/tasks/release-checks/tests/generic/au-cypress.js (hunk 2 above)
Comment (Sayan751) [reply build/tasks/release-checks/tests/generic/au-cypress.js]: Absolutely. Thanks for pointing out. Will correct this.

File: skeleton/common/tsconfig.json__if_typescript (hunk 1 above)
Comment (3cp) [reply skeleton/common/tsconfig.json__if_typescript]: errr... `import * as json from 'a.json';` does not work with `"esModuleInterop": false`?

File: skeleton/webpack/webpack.config.js (hunk 3 above)
Comment (Sayan751) [reply skeleton/webpack/webpack.config.js]: My earlier take on the typeRoots involving monorepo might be little bit premature, as I have not recently tried to create a monorepo for a aurelia project. I tried that that a year back, and there were massive problems with resolving/loading symlinked packages; the webpack config was crowded with `alias`es. If this is a blocker, I will try to create a lerna prototype, and report the the findings.

**Comment by Sayan751** [comment]
> One more question, I didn't see you remove aurelia-cli from final package.json of webpack app. Is unit tests still using `au test` command?

Yup, those are still based on `au test`. But it can also be encapsulated in a npm script. As I was focusing on webpack, I actually paid no attention to that.

**Comment by 3cp** [comment]
I thought our target is to remove runtime dependency on aurelia-cli for webpack app.

File: skeleton/common/tsconfig.json__if_typescript (hunk 1 above)
Comment (Sayan751) [reply skeleton/common/tsconfig.json__if_typescript]: I started with `resolveJsonModule` only, but it didn't work alone. Thereafter I use both of them together (muscle memory 😛). However, this info is also quite a few months old. Not sure, if it works without `esModuleInterop`.

Comment (Sayan751) [reply skeleton/common/tsconfig.json__if_typescript]: I found the issue to be closed: https://github.com/Microsoft/TypeScript/issues/25400

I will try without `esModuleInterop`.

**Comment by Sayan751** [comment]
> I thought our target is to remove runtime dependency on aurelia-cli for webpack app.

I misunderstood it to be a task to remove the `au run`, and `au build` for webpack projects. It makes more sense to remove all `au` flavored commands from webpack apps. I would suggest to have that in a different PR, as this is already quite big. A new PR might also be easy to review. We can still use the same issue to track that.

**Comment by 3cp** [comment]
No rush, we appreciate your work on this. As this is likely a post-v1 improvement, you have plenty of time.

**Comment by 3cp** [comment]
BTW, please follow the commit message convention to be properly logged in changelog and release note. https://github.com/DurandalProject/about/blob/master/CONTRIBUTING.md#commit

**Comment by Sayan751** [comment]
Missed the git commit message guideline clearly. Will adhere to that in following commits.

**Comment by 3cp** [comment]
You can squash all commits into one if you want, that can avoid some similar entries from multiple commits in changelog.

File: skeleton/common/tsconfig.json__if_typescript (hunk 1 above)
Comment (Sayan751) [reply skeleton/common/tsconfig.json__if_typescript]: Removed the `esModuleInterop`. The JSON import is working corrctly.

File: build/tasks/release-checks/tests/generic/au-cypress.js (hunk 2 above)
Comment (Sayan751) [reply build/tasks/release-checks/tests/generic/au-cypress.js]: Correction done

File: skeleton/webpack/webpack.config.js (hunk 3 above)
Comment (Sayan751) [reply skeleton/webpack/webpack.config.js]: I tried to create a [monorepo using au-cli and lerna](https://github.com/Sayan751/au-cli-lerna). I couldn't make it work. It is not really because of the type definition, which makes the problem I suspected earlier. But even after fixing the typeRoots, Aurelia has problems loading the custom elements coming from different plugins. I opened an [issue](https://github.com/aurelia/webpack-plugin/issues/145) before on similar problem. I think the problems related to loading the symlinked packages needs to corrected before starting with bootstrapping lerna monorepo projects with lerna. 

On the other hand, this typing issue with Jasmine is already opened here: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/14569. If type definition  of jasmine is got resolved, probably the whole `typeRoots` thing can be done away with.

**Comment by Sayan751** [comment]
@3cp @fkleuver It is ready for review again.

**Comment by Sayan751** [comment]
Just wanted to point out that this pull request sort of addresses #1039 as well.

File: skeleton/webpack/webpack.config.js (hunk 3 above)
Comment (3cp) [reply skeleton/webpack/webpack.config.js]: I also discovered another bug with limiting ts-loader include. I am running release-check on v1 candidate, where I need to modify it to `include: karma ? [srcDir, testDir] : srcDir` to fix webpack+karma setup.

Comment (Sayan751) [reply skeleton/webpack/webpack.config.js]: Yes, I faced that too. Otherwise the `*.spec.ts` files are never processed, and webpack throws error. I think not having the `include` for `ts-loader` is far easier solution.

Comment (3cp) [reply skeleton/webpack/webpack.config.js]: I see. I will retain the "include" for v1 release. Then we can work on your big refactoring after v1.

**Comment by 3cp** [comment]
FYI, it's not `<feat>(topic):`, it's just `feat(topic):`.

**Comment by alexdresko** [comment]
*bump*

**Comment by fkleuver** [comment]
@Sayan751 can you rewrite that one commit message with feat to follow conventional commit guidelines please? only then will it correctly show up in generated changelogs
I can't really comment much on the contents, @3cp what are your thoughts at this point?

**Comment by 3cp** [comment]
We need to polish it up, and have some kind of plan to handle this kind of breaking change.

We (Sayan included) are all bit busy with Aurelia 2 😅 , not spending enough energy here.

**Comment by Sayan751** [comment]
@fkleuver @3cp I can try to squash all the commits to one with conventional commit message and force push, if that is ok?

**Comment by 3cp** [comment]
Force push is fine.

**Comment by Sayan751** [comment]
@fkleuver @3cp I tried to squash all the changes, but there are many conflicting changes. These changes have got pretty old and the fork got also out of sync. I think at this point it is best to create a new PR. I have started another branch that contains all the changes squashed into one commit that includes your suggestions. That branch is also synced with the Aurelia master. Currently I am running the release checks, once that is done, I would like to create a new PR and close this one. Let me if you think otherwise.

**Comment by Sayan751** [comment]
Closed in favor of #1137

**Comment by alexdresko** [comment]
So can this be used now?

**Comment by Sayan751** [comment]
Sorry @alexdresko you need to wait till #1137 is merged and published.

**Comment by alexdresko** [comment]
@Sayan751  Okay _now_ can we use this? :)
